Add ADR-04 about who writes files, and when (#5)
ethersync/ethersync: docs/decisions/04-file-saving.md

The question is: Who writes to files, and when? Who has "ownership" of the files?

## Decision Drivers

* It should feel as natural as possible to work in a shared directory.
* It should be possible to work on the files using external tools, like `awk`, or appending things to the files.

## Considered Options

When a file has been opened in editors:

* The daemon keeps an opened file content up-to-date with the CRDT content.
* When editors open a file, they take ownership of it. The daemon doesn't write the file.

When no editor has the file open:

* The daemon keeps the files up-to-date with its CRDT content immediately
* The daemon writes the current state occasionally, with its own caching/writing logic
* The daemon never writes to the files

## Decision Outcome

Chosen option: "When editors open a file, they take ownership of it. When no editor has the file open, the daemon writes to it using its own logic"

Notes:

- We could provide another "interface" to the content, through ethersync tooling, like `ethersync run "sed -i s/foo/bar file"`.
- Maybe we should implement a ping mechanism, so that the daemon can detect when editors have crashed.

> - When a file is opened in more than one editor at once, there will be conflicting writes anyway. This is true without a software like Ethersync. The only way to prevent that would be to stop editors from writing at all, which doesn't seem desirable.

Reviewer: ```suggestion
- When a file is opened in more than one editor at once, there could be conflicting writes anyway. This is true without a software like Ethersync. The only way to prevent that would be to stop editors from writing at all, which doesn't seem desirable.
```

> - To incorporate changes done by external tools, we watch the file system, compute diffs, and apply changes to the CRDT. This seems true both when the file is open or closed in editors.

Reviewer: ```suggestion
- To incorporate changes done by external tools, we watch the file system, compute diffs, and apply changes to the CRDT. This is independent of whether the file is open or closed in editors.
```
